File: client/src/utils/validateEmail.js
export function validateEmail(email) {
  const SPECIAL_CHARS = ['!', '#', '$', '%', '&', '\'', '*', '+', '-', '/', '=', '?', '^', '_', '`', '{', '|', '}', '~', '.', ','];
  const ALLOWED_CHARS = ['-', '.'] // allowed special characters
  const MAX_LOCAL_LENGTH = 64;
  const MAX_DOMAIN_LENGTH = 255;
  const MAX_SUBDOMAIN_LENGTH = 63;
  const MAX_LENGTH = MAX_LOCAL_LENGTH + MAX_DOMAIN_LENGTH + 1; // +1 for @

  // empty
  if (email === '' || email === null || email === undefined) {
    return { valid: false, message: 'empty' };
  }
  
  // does not contain @
  if (email.includes('@') === false) {
    return { valid: false, message: 'no @' };
  }

  // contains space
  if (email.includes(' ') === true) {
    return { valid: false, message: 'space' };
  }

  // contains multiple @
  if (email.split('@').length > 2) {
    return { valid: false, message: 'multiple @' };
  }

  // empty or too long
  if (email.length < 3 || email.length > MAX_LENGTH) {
    return { valid: false, message: 'empty or too long' };
  }

  let [local, domain] = email.split('@');

  // LOCAL VALIDATION
  // local part is empty or too long
  if (local.length < 1 || local.length > MAX_LOCAL_LENGTH) {
    return { valid: false, message: 'local empty or too long' };
  }

  // has double special characters
  for (let i = 0; i < local.length; i++) {
    if (SPECIAL_CHARS.includes(local[i]) && SPECIAL_CHARS.includes(local[i + 1])) {
      return { valid: false, message: 'local has double special characters' };
    }
  }

  // has special character at beginning or end
  if (SPECIAL_CHARS.includes(local[0]) || SPECIAL_CHARS.includes(local[local.length - 1])) {
    return { valid: false, message: 'local starts or ends with special character' };
  } 

  // DOMAIN VALIDATION
  // domain part is empty or too long
  if (domain.length < 1 || domain.length > MAX_DOMAIN_LENGTH) {
    return { valid: false, message: 'domain empty or too long' };
  }

  // domain part contains special characters
  const charFilter = function (char) { // filters allowed special characters out of disallowed ones
    return !ALLOWED_CHARS.includes(char);
  }
  if (SPECIAL_CHARS.filter(charFilter).some(char => domain.includes(char))) {
    return { valid: false, message: 'domain contains special characters' };
  }

  // CHECKS SUBDOMAINS
  let domains = domain.split('.');

  // must have at least 2 subdomains
  if (domains.length < 2) {
    return { valid: false, message: 'must have at least 2 subdomains' };
  }

  // domain part too small or too large
  for (let i = 0; i < domains.length; i++) {
    if (domains[i].length < 1 || domains[i].length > MAX_SUBDOMAIN_LENGTH) {;
      return { valid: false, message: 'subdomain too small or too large' };
    }

    // starts or ends with -
    if (domains[i][0] === '-' || domains[i][domains[i].length - 1] === '-') {
      return { valid: false, message: 'subdomain starts or ends with -' };
    }
  }

  let tld = domains[domains.length - 1];
  // tld too small
  if (tld.length < 2) {
    return { valid: false, message: 'tld too small' };
  }

  // passes checks
  return { valid: true, message: 'passes checks'};
}

File: server/core/utils/validate_email.py
def validate_email(email):
    SPECIAL_CHARS = ['!', '#', '$', '%', '&', '\'', '*', '+', '-', '/', '=', '?', '^', '_', '`', '{', '|', '}', '~', '.', ',']
    ALLOWED_CHARS = ['.', '-'] # allowed characters in domain portion
    MAX_LOCAL_LENGTH = 64
    MAX_DOMAIN_LENGTH = 255
    MAX_SUBDOMAIN_LENGTH = 63
    MAX_LENGTH = MAX_LOCAL_LENGTH + MAX_DOMAIN_LENGTH + 1 # +1 for '@'

    # empty
    if not email:
        return False
    
    # does not contain '@'
    if '@' not in email:
        return False
    
    # contains space
    if ' ' in email:
        return False
    
    # contains multiple '@'
    if email.count('@') > 1:
        return False
    
    # too short or too long
    if len(email) < 3 or len(email) > MAX_LENGTH:
        return False
    
    local, domain = email.split('@')

    # LOCAL VALIDATION
    # local is empty or too long
    if not local or len(local) > MAX_LOCAL_LENGTH:
        return False
    
    # has double special characters
    for special in SPECIAL_CHARS:
        if special * 2 in local:
            return False
        
    # has special character at beginning or end
    if local[0] in SPECIAL_CHARS or local[-1] in SPECIAL_CHARS:
        return False
    
    # DOMAIN VALIDATION
    # domain is empty or too long
    if not domain or len(domain) > MAX_DOMAIN_LENGTH:
        return False
    
    # has special characters
    for special in SPECIAL_CHARS:
        if special in ALLOWED_CHARS:
            continue

        if special in domain:
            return False
        
    # SUBDOMAIN VALIDATION
    subdomains = domain.split('.')

    # must have at least two subdomains (domain and tld)
    if len(subdomains) < 2:
        return False
    
    for sub in subdomains:
        # subdomain is too short or too long
        if len(sub) < 1 or len(sub) > MAX_SUBDOMAIN_LENGTH:
            return False
        
        # starts or ends with an ALLOWED_CHAR
        if sub[0] in ALLOWED_CHARS or sub[-1] in ALLOWED_CHARS:
            return False
        
    # TLD VALIDATION
    tld = subdomains[-1]

    # tld is too short
    if len(tld) < 2:
        return False
    
    return True
        


